File: Generecidad/Ejer3/Pila.py
# 5: Crea una clase genérica Pila<T>
class Pila:
    def __init__(self):
        self.elementos = []
#a) Implementa un método para apilar
    def apilar(self, elemento):
        self.elementos.append(elemento)
#b) Implementa un método para des apilar
    def desapilar(self):
        if self.elementos:
            return self.elementos.pop()
        print("La pila está vacía")
        return None

    def mostrar_pila(self):
        print("Contenido de la pila:", self.elementos)

#c) Prueba la pila con diferentes tipos de datos
#d) Muestra los datos de la pila
pila_enteros = Pila()
pila_enteros.apilar(10)
pila_enteros.apilar(20)
pila_enteros.mostrar_pila()
print("Elemento desapilado:", pila_enteros.desapilar())
pila_enteros.mostrar_pila()

pila_strings = Pila()
pila_strings.apilar("Python")
pila_strings.apilar("Java")
pila_strings.mostrar_pila()
print("Elemento desapilado:", pila_strings.desapilar())
pila_strings.mostrar_pila()

File: Generecidad/Ejer3/Pila.java
import java.util.Stack;

//5: Crea una clase genérica Pila<T>
public class Pila<T> {
    private Stack<T> elementos;

    public Pila() {
        this.elementos = new Stack<>();
    }

    // a) Implementa un método para apilar
    public void apilar(T elemento) {
        elementos.push(elemento);
    }

    // b) Implementa un método para des apilar
    public T desapilar() {
        if (!elementos.isEmpty()) {
            return elementos.pop();
        }
        System.out.println("La pila está vacía");
        return null;
    }

    public void mostrarPila() {
        System.out.println("Contenido de la pila: " + elementos);
    }

    public static void main(String[] args) {
        // c) Prueba la pila con diferentes tipos de datos
        // d) Muestra los datos de la pila
        Pila<Integer> pilaEnteros = new Pila<>();
        pilaEnteros.apilar(10);
        pilaEnteros.apilar(20);
        pilaEnteros.mostrarPila();
        System.out.println("Elemento desapilado: " + pilaEnteros.desapilar());
        pilaEnteros.mostrarPila();

        Pila<String> pilaStrings = new Pila<>();
        pilaStrings.apilar("Python");
        pilaStrings.apilar("Java");
        pilaStrings.mostrarPila();
        System.out.println("Elemento desapilado: " + pilaStrings.desapilar());
        pilaStrings.mostrarPila();
    }
}